File: sdks/python/tests/blackbox.py
"""Black-box Python SDK test against an independently running real broker."""

import json
import os
from pathlib import Path

from abyss_sdk import BrokerApiError, BrokerClient
from abyss_sdk.plugin import AbyssPlugin


def main() -> None:
    startup_info = os.environ.get("ABYSS_BROKER_STARTUP_INFO")
    if not startup_info:
        raise RuntimeError("ABYSS_BROKER_STARTUP_INFO must point to a real broker")

    client = BrokerClient.from_startup_info(startup_info)
    startup = json.loads(Path(startup_info).read_text(encoding="utf-8"))
    unauthenticated = BrokerClient(base_url=f"http://{startup['api_addr']}")
    events = AbyssPlugin(plugin_id="blackbox.python-sdk").connect()

    assert client.get_health() == {"service": "abyss-broker", "status": "ok"}
    try:
        unauthenticated.get_mitm_config()
    except BrokerApiError as error:
        assert error.status == 401
    else:
        raise AssertionError("protected broker REST route accepted no bearer token")
    assert client.get_proxy_status()["lifecycle"] == "running"
    mitm = client.get_mitm_config()
    assert client.update_mitm_config(mitm) == mitm
    hooks = client.get_hooks_config()
    hooks["harness_usage"]["config"]["harnesses"]["python-sdk-custom"] = {
        "enabled": True,
        "matchers": [{"process_names": ["python-sdk-custom"]}],
    }
    updated_hooks = client.update_hooks_config(hooks)
    assert (
        updated_hooks["harness_usage"]["config"]["harnesses"]["python-sdk-custom"]
        == hooks["harness_usage"]["config"]["harnesses"]["python-sdk-custom"]
    )
    client.collect_broker_logs({"max_bytes_per_file": 4096})
    assert client.get_diagnostics()["schema_version"] == 1
    assert client.get_network_observations(10)["schema_version"] == 1
    client.get_traffic_snapshot()

    assert client.shutdown()["lifecycle"] == "stopped"
    assert list(events) == []
    assert events.close is not None and events.close.code == 100
    print("Python SDK real-broker black-box: ok")


if __name__ == "__main__":
    main()

File: sdks/typescript/test/blackbox.mjs
import assert from "node:assert/strict";
import { readFile } from "node:fs/promises";

import { BrokerApiError, BrokerClient } from "../dist/index.js";
import { AbyssPlugin } from "../dist/plugin/index.js";

const startupInfo = process.env.ABYSS_BROKER_STARTUP_INFO;
if (!startupInfo) {
  throw new Error("ABYSS_BROKER_STARTUP_INFO must point to a real broker");
}

const client = await BrokerClient.fromStartupInfo(startupInfo);
const startup = JSON.parse(await readFile(startupInfo, "utf8"));
const unauthenticated = new BrokerClient({
  baseUrl: `http://${startup.api_addr}`,
});
const events = await new AbyssPlugin({
  consumerId: "blackbox.typescript-sdk",
}).connect();

assert.deepEqual(await client.getHealth(), {
  service: "abyss-broker",
  status: "ok",
});
await assert.rejects(
  unauthenticated.getMitmConfig(),
  (error) => error instanceof BrokerApiError && error.status === 401,
);
assert.equal((await client.getProxyStatus()).lifecycle, "running");

const mitm = await client.getMitmConfig();
assert.deepEqual(await client.updateMitmConfig(mitm), mitm);
const hooks = await client.getHooksConfig();
hooks.harness_usage.config.harnesses["typescript-sdk-custom"] = {
  enabled: true,
  matchers: [{ process_names: ["typescript-sdk-custom"] }],
};
const updatedHooks = await client.updateHooksConfig(hooks);
assert.deepEqual(
  updatedHooks.harness_usage.config.harnesses["typescript-sdk-custom"],
  hooks.harness_usage.config.harnesses["typescript-sdk-custom"],
);
await client.collectBrokerLogs({ max_bytes_per_file: 4096 });
assert.equal((await client.getDiagnostics()).schema_version, 1);
assert.equal((await client.getNetworkObservations(10)).schema_version, 1);
await client.getTrafficSnapshot();

assert.equal((await client.shutdown()).lifecycle, "stopped");
assert.equal(await events.nextEvent(), undefined);
assert.equal(events.close?.code, 100);

process.stdout.write("TypeScript SDK real-broker black-box: ok\n");
